crates/nu-lsp/src/symbols.rs

```diff
@@ -0,0 +1,534 @@
+    }
+}
+
+impl PartialOrd for Symbol {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Symbol {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.kind == other.kind {
+            return self.range.start.cmp(&other.range.start);
+        }
+        match (self.kind, other.kind) {
+            (SymbolKind::FUNCTION, _) => Ordering::Less,
+            (_, SymbolKind::FUNCTION) => Ordering::Greater,
+            _ => self.range.start.cmp(&other.range.start),
+        }
+    }
+}
+
+impl Symbol {
+    fn to_symbol_information(&self, uri: &Uri) -> SymbolInformation {
+        #[allow(deprecated)]
+        SymbolInformation {
+            location: Location {
+                uri: uri.clone(),
+                range: self.range,
+            },
+            name: self.name.to_owned(),
+            kind: self.kind,
+            container_name: None,
+            deprecated: None,
+            tags: None,
+        }
+    }
+}
+
+/// Cache symbols for each opened file to avoid repeated parsing
+pub struct SymbolCache {
+    /// Fuzzy matcher for symbol names
+    matcher: SkimMatcherV2,
+    /// File Uri --> Symbols
+    cache: BTreeMap<Uri, Vec<Symbol>>,
+    /// If marked as dirty, parse on next request
+    dirty_flags: BTreeMap<Uri, bool>,
+}
+
+impl SymbolCache {
+    pub fn new() -> Self {
+        SymbolCache {
+            matcher: SkimMatcherV2::default(),
+            cache: BTreeMap::new(),
+            dirty_flags: BTreeMap::new(),
+        }
+    }
+
+    pub fn mark_dirty(&mut self, uri: Uri, flag: bool) {
+        self.dirty_flags.insert(uri, flag);
+    }
+
+    fn get_symbol_by_id(
+        working_set: &StateWorkingSet,
+        id: Id,
+        doc: &FullTextDocument,
+        doc_span: &Span,
+    ) -> Option<Symbol> {
+        match id {
+            Id::Declaration(decl_id) => {
+                let decl = working_set.get_decl(decl_id);
+                let span = working_set.get_block(decl.block_id()?).span?;
+                // multi-doc working_set, returns None if the Id is in other files
+                if !doc_span.contains(span.start) {
+                    return None;
+                }
+                Some(Symbol {
+                    name: decl.name().to_string(),
+                    kind: SymbolKind::FUNCTION,
+                    range: span_to_range(&span, doc, doc_span.start),
+                })
+            }
+            Id::Variable(var_id) => {
+                let var = working_set.get_variable(var_id);
+                let span = var.declaration_span;
+                if !doc_span.contains(span.start) || span.end == span.start {
+                    return None;
+                }
+                let range = span_to_range(&span, doc, doc_span.start);
+                let name = doc.get_content(Some(range));
+                // TODO: better way to filter closures with type any
```

Comment:
I don't know if it's supposed to return variables whose span covers an entire block like `{|xx| xxxx}`